keps/sig-instrumentation/1933-secret-logging-static-analysis/README.md

```diff
@@ -0,0 +1,608 @@
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+While the test will not initially be blocking, if it produces many false positives, 
+it may train developers to ignore issues.
+Additionally, *because* it is not initially blocking, it risks being overlooked
+as unimportant even when findings are relevant.
+
+Should issues reach master, either by the happenstance of a merge, ignored warnings, or analysis flakiness,
+reported findings may include those out of scope for the change set in a given PR.
+Such incidents are little more than confusing toil for developers.
+
+Changes to `test-infra` carry with them the potential for inconvenience,
+should they introduce any instability to wider testing.  While diligent review
+mitigates this, it does not remove the concern completely.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+In-depth analysis interacts with build artifacts, and as a result is quickly
+stymied by Bazel's sandboxing of build and test environments.
+It is similar in this regard to linters, and indeed a standard execution path for
+analysis in Go is via `go vet -vettool=...`.
+
+This KEP would introduce a execution script to `hack/` which handles passing
+the analyzer and config to our vet process.
+Within Kubernetes, `go vet` is executed via `make vet`.
+The analyzer can be passed to this in the introduced `hack/` script via
+`make vet WHAT="..."`.
+Dependency on `go-flow-levee` will be introduced to `hack/tools` to avoid polluting
+the main Kubernetes dependencies.
+
+This KEP will also introduce execution of this script as a Prow presubmit job.
+Any findings by the analysis will constitute a test failure.
+Failure will be initially non-blocking, per the graduation criteria below.
+Analysis findings reporting the position of both source and sink,
+ indicating to a developer which log call must be corrected and/or which argument must be sanitized.
+
+### Test Plan
+
+As a testing target, testing will consist of examples of Kubernetes-specific
+cases that we expect static analysis to detect.  No full-scale Kubernetes
+integration/e2e tests will be necessary.
+
+For analysis based on `go-flow-levee`, this will consist of sample Kubernetes code based on the same configuration used in presubmit scanning.
+This test will be examined via the `analysistest` package to ensure analysis produces expected diagnostics.
+As part of testing of our testing process, these tests should belong to `kubernetes/test-infra`.
+
+### Graduation Criteria
+
+#### Alpha (1.20)
+- Analysis runs as a non-blocking presubmit check, warning developers of any findings in their changes.
```

Comment:
Ideally it should probably go through these phases:
- manually run or non-reporting presubmit
  - validate that this is running soundly
- reporting, non-blocking presubmit
  - validate _at scale_ that this is running soundly
- reporting, blocking presubmit
  - this phase can only happen if it's quick (seems likely) and stable / reliable (seems likely, but we should confirm)

we have found that many do not understand "reporting, but not required / blocking" so usually we want that to only be a short period (perhaps a week or two) where we are pretty confident in the job but still want to dry run making it blocking.

because prowjobs can be tricky, it should be manually triggered or not reporting when initially introduced, just long enough to validate the job for a few runs. it can move to reporting non-blocking quickly.

we don't add new presubmits often, generally encouraging things to move to postsubmit, but I think this is more than reasonable here from what I've seen so far ...

Reply:
That all sounds good to me.

There is some work to be done yet, mostly covered by KEP-1753, in the identification of what fields should not be logged.  In local testing, we've been using an otherwise pretty ad hoc configuration for that.  If the `beta` stage is going to be short lived, it might be best to stay in `alpha` until that information can be consumed.  I've moved that bullet point to the `alpha -> beta` graduation.